bpf/lib/nodeport.h

```diff
@@ -1203,6 +1319,90 @@ static __always_inline int xlate_dsr_v4(struct __ctx_buff *ctx,
 	return ret;
 }
 
+static __always_inline int dsr_reply_icmp4(struct __ctx_buff *ctx,
+					   struct iphdr *ip4 __maybe_unused,
+					   int code, int ohead __maybe_unused)
+{
+#ifdef ENABLE_DSR_ICMP_ERRORS
+	const __s32 orig_dgram = 8, off = ETH_HLEN;
+	const __u32 l3_max = MAX_IPOPTLEN + sizeof(*ip4) + orig_dgram;
+	__be16 type = bpf_htons(ETH_P_IP);
+	__s32 len_new = off + ipv4_hdrlen(ip4) + orig_dgram;
+	__s32 len_old = ctx_full_len(ctx);
+	__u8 tmp[l3_max];
+	union macaddr smac, dmac;
+	struct icmphdr icmp __align_stack_8 = {
+		.type		= ICMP_DEST_UNREACH,
+		.code		= ICMP_FRAG_NEEDED,
+		.un = {
+			.frag = {
+				.mtu = bpf_htons(THIS_MTU - ohead),
+			},
+		},
+	};
+	struct iphdr ip __align_stack_8 = {
+		.ihl		= sizeof(ip) >> 2,
+		.version	= IPVERSION,
+		.ttl		= IPDEFTTL,
+		.tos		= ip4->tos,
+		.id		= ip4->id,
+		.protocol	= IPPROTO_ICMP,
+		.saddr		= ip4->daddr,
+		.daddr		= ip4->saddr,
+		.frag_off	= bpf_htons(IP_DF),
+		.tot_len	= bpf_htons(sizeof(ip) + sizeof(icmp) +
```

Comment:
Is this correct? `len_new` is assigned to `off + ipv4_hdrlen(ip4) + orig_dgram` so here we are setting `tot_len` to 

```
sizeof(ip) + sizeof(icmp) + off + ipv4_hdrlen(ip4) + orig_dgram - off
```

Reply:
Yes, it's correct. The above is setting `sizeof(ip) + sizeof(icmp) + len_new - off`, in other words this is: IP + ICMP + IP + 8b payload.